src/goto-symex/symex_complexity.h

```diff
@@ -0,0 +1,139 @@
+/*******************************************************************\
+
+Module: Symbolic Execution
+
+Author: John Dumbell
+
+\*******************************************************************/
+
+/// \file
+/// Symbolic Execution
+
+#ifndef CPROVER_GOTO_SYMEX_SYMEX_COMPLEXITY_H
+#define CPROVER_GOTO_SYMEX_SYMEX_COMPLEXITY_H
+
+#include "complexity_violation.h"
+#include "goto_symex_state.h"
+#include "symex_heuristic_transformation.h"
+
+/// Wrapper class for active loop information.
+class loop_complexity_infot
+{
+public:
+  loop_complexity_infot()
+  {
+  }
+
+  loop_complexity_infot(irep_idt _loop_id, framet::loop_infot *_loop_info)
+    : loop_id(_loop_id), loop_info(_loop_info)
+  {
+  }
+
+  bool empty()
+  {
+    return loop_id.empty() && loop_info == nullptr;
+  }
+
+  irep_idt loop_id;
+  framet::loop_infot *loop_info = nullptr;
+};
+
+/// Symex complexity module.
+///
+/// Dynamically cancels branches when symex considers them too complex.
+/// A branch is too complex when runtime will be beyond reasonable
+/// bounds due to internal structures becoming too large or solving conditions
+/// becoming too big for the SAT solver to deal with.
+///
+/// On top of the branch cancelling there is special consideration for loops.
+/// As branches get cancelled it keeps track of what loops are currently active
+/// up through the stack. If a loop has too many of its child branches killed
+/// (including additional loops or recursion) it won't do any more unwinds of
+/// that particular loop and blacklist it.
+///
+/// This blacklist is only held for as long as any loop is still active. As soon
+/// as the loop iteration ends and the context in which the code is being
+/// executed changes it'll be able to be run again.
+///
+/// Example of loop blacklisting:
+///
+///     loop A: (complexity: 5070)
+///         loop B: (complexity: 5000)
+///         loop C: (complexity: 70)
+///
+/// In the above loop B will be blacklisted if we have a complexity limitation
+/// < 5000, but loop A and B will still be run, because when loop B is removed
+/// the complexity of the loop as a whole is acceptable.
+class symex_complexityt
+{
+public:
+  symex_complexityt() = default;
+
+  explicit symex_complexityt(message_handlert &logger) : log(logger)
+  {
+  }
+
+  symex_complexityt(message_handlert &logger, const optionst &options);
+
+  /// Is the complexity module active?
+  /// \return Whether complexity limits are active or not.
+  inline bool complexity_limits_active()
+  {
+    return this->complexity_active;
+  }
+
+  /// Checks the passed-in state to see if its become too complex for us to deal
+  /// with, and if so set its guard to false.
+  /// \param state goto_symex_statet you want to check the complexity of.
+  /// \return True/false depending on whether this branch has been abandoned.
```

Comment:
has been -> should be? Or does this actually do the work itself (in which case rename accordingly, something like `truncate_path_if_too_complex`; at present the name suggests a const method